File: 2023/day12.php
<?php

/**
 * Sadly this isn't all mine, mostly taken from python example and ported to php I tried so hard... :'(
 */

require 'common.php';

$lines = get_input(12, false);



$sum = 0;
$sum2 = 0;
foreach ($lines as $line) {
    [$s, $g] = explode(' ', $line);
    $groups = explode(',', $g);
    $springs = $s . '.';

    $groups2 = explode(',', rtrim(str_repeat($g . ',', 5), ','));
    $springs5 = str_repeat($s . '?', 5);
    $springs2 = substr($springs5, 0, strlen($springs5) - 1) . '.';
    // echo $springs . PHP_EOL . implode(',', $groups) . PHP_EOL;
    echo " ----- new group ----- " . PHP_EOL;

    $cache = [];
    $s = count_patterns(0, 0, $springs, $groups);
    echo "SUM for part 1 group: $s" . PHP_EOL;
    $sum += $s;

    $cache = [];
    // echo "$springs2 - " . implode(',', $groups2) . PHP_EOL;
    $s2 = count_patterns(0, 0, $springs2, $groups2);
    echo "SUM for part 2 group: $s2" . PHP_EOL;
    $sum2 += $s2;
}

d($sum);
d($sum2);

function count_patterns($spring_pos, $group_num, $springs, $groups)
{
    global $cache;
    // echo "Starting Count " . $spring_pos . ' ' . $group_num . PHP_EOL;
    if ($group_num >= count($groups)) {
        if ($spring_pos < strlen($springs) && str_contains(substr($springs, $spring_pos), '#')) {
            return 0; // not a solution
        }
        return 1;
    }

    if ($spring_pos >= strlen($springs)) {
        return 0; // ran out of springs
    }

    if (isset($cache["$spring_pos,$group_num"])) {
        return $cache["$spring_pos,$group_num"];
    }

    $count = 0;
    $group_size = $groups[$group_num];
    //$current_group = substr($springs, $spring_pos, $group_size);
    $current_spring = $springs[$spring_pos];
    //$last_spring = $springs[$spring_pos + $group_size];

    // echo 'Starting: ' . $spring_pos . ' ' . $group_num . ' ' . $group_size . PHP_EOL;
    if ($current_spring === '?') {
        //echo 'Current group: ' . substr($springs, $spring_pos, $group_size) . ' last spring: ' . $springs[$spring_pos + $group_size] . PHP_EOL;
        if (! str_contains(substr($springs, $spring_pos, $group_size), '.') && $springs[$spring_pos + $group_size] !== '#') {
            $count = count_patterns($spring_pos + $group_size + 1, $group_num + 1, $springs, $groups) + count_patterns($spring_pos + 1, $group_num, $springs, $groups);
        } else {
            $count = count_patterns($spring_pos + 1, $group_num, $springs, $groups);
        }
    } elseif ($current_spring === '#') {
        if (! str_contains(substr($springs, $spring_pos, $group_size), '.') && $springs[$spring_pos + $group_size] !== '#') {
            $count = count_patterns($spring_pos + $group_size + 1, $group_num + 1, $springs, $groups);
        } else {
            $count = 0;
        }
    } elseif ($current_spring === '.') {
        $count = count_patterns($spring_pos + 1, $group_num, $springs, $groups);
    }

    $cache["$spring_pos,$group_num"] = $count;
    //echo $spring_pos . ' ' . $group_num . ' ' . $group_size . ' ' . $current_group . ' ' . $current_spring . ' ' . $last_spring . ' ' . $count . PHP_EOL;
    return $count;
}

File: 2023/day12.py
def arrangement_count(p, g, springs, groups):
    #print('Starting Count ' + str(p) + ' ' + str(g))
    if g >= len(groups): # no more groups
        if p < len(springs) and '#' in springs[p:]:
            # eg: .##?????#.. 4,1
            return 0 # not a solution - there are still damaged springs in the record
        return 1
    
    if p >= len(springs):
        return 0 # we ran out of springs but there are still groups to arrange

    res = None
    gs = groups[g] # damaged group size

    print(f"Starting {p} {g} {gs}")
    if springs[p] == '?':
        # if we can start group of damaged springs here
        # eg: '??#...... 3' we can place 3 '#' and there is '?' or '.' after the group
        # eg: '??##...... 3' we cannot place 3 '#' here
        #print(f"current_group: {springs[p:p+gs]} last spring: {springs[p+gs]}")
        if '.' not in springs[p:p + gs] and springs[p + gs] != '#':
            # start damaged group here + this spring is operational ('.')
            res = arrangement_count(p + gs + 1, g + 1, springs, groups) + arrangement_count(p + 1, g, springs, groups)
            print('A', res)
        else:
            # this spring is operational ('.')
            res = arrangement_count(p + 1, g, springs, groups)
            print('B', res)
    elif springs[p] == '#':
        # if we can start damaged group here
        #print(f"current_group: {springs[p:p+gs]} last spring: {springs[p+gs]}")
        if '.' not in springs[p:p + gs] and springs[p + gs] != '#':
            res = arrangement_count(p + gs + 1, g + 1, springs, groups)
            print('C', res)
        else:
            res = 0 # not a solution - we must always start damaged group here
            print('D', res)
    elif springs[p] == '.':
        res = arrangement_count(p+1, g, springs, groups) # operational spring -> go to the next spring
        print('E', res)

    return res

with open('day12.ex') as f:
    sum_of_arrangements = 0

    for line in f.readlines():
        springs, groups = line.split()

        groups = list(map(int, groups.split(',')))
        springs = springs + '.' # make sure there is operational spring after each damaged group

        # print(springs)
        # print(groups)
        print(" ----- new group ----- ")
        sum = arrangement_count(0, 0, springs, groups)
        print(f"Sum for group: {sum}")
        sum_of_arrangements += sum

    print(sum_of_arrangements)